homeassistant/components/modern_forms/config_flow.py

```diff
@@ -0,0 +1,118 @@
+from homeassistant.helpers.typing import DiscoveryInfoType
+
+from .const import DOMAIN
+
+
+class ModernFormsFlowHandler(ConfigFlow, domain=DOMAIN):
+    """Handle a ModernForms config flow."""
+
+    VERSION = 1
+    CONNECTION_CLASS = CONN_CLASS_LOCAL_POLL
+
+    async def async_step_user(
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Handle setup by user for Modern Forms integration."""
+        return await self._handle_config_flow(user_input)
+
+    async def async_step_zeroconf(
+        self, discovery_info: DiscoveryInfoType
+    ) -> FlowResult:
+        """Handle zeroconf discovery."""
+        host = discovery_info["hostname"].rstrip(".")
+        name, _ = host.rsplit(".")
+
+        self.context.update(
+            {
+                CONF_HOST: discovery_info["host"],
+                CONF_NAME: name,
+                CONF_MAC: discovery_info["properties"].get(CONF_MAC),
+                "title_placeholders": {"name": name},
+            }
+        )
+
+        # Prepare configuration flow
+        return await self._handle_config_flow(discovery_info, True)
+
+    async def async_step_zeroconf_confirm(
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Handle a flow initiated by zeroconf."""
+        return await self._handle_config_flow(user_input)
+
+    async def _handle_config_flow(
+        self, user_input: dict[str, Any] | None = None, prepare: bool = False
+    ) -> FlowResult:
+        """Config flow handler for ModernForms."""
+        source = self.context.get("source")
+
+        # Request user input, unless we are preparing discovery flow
+        if user_input is None and not prepare:
+            if source == SOURCE_ZEROCONF:
+                return self._show_confirm_dialog()
+            return self._show_setup_form()
+
+        if source == SOURCE_ZEROCONF:
+            user_input[CONF_HOST] = self.context.get(CONF_HOST)  # type: ignore
+            user_input[CONF_MAC] = self.context.get(CONF_MAC)  # type: ignore
+
+        if user_input.get(CONF_MAC) is None or not prepare:  # type: ignore
+            session = async_get_clientsession(self.hass)
+            device = ModernFormsDevice(user_input[CONF_HOST], session=session)  # type: ignore
+            try:
+                device = await device.update()
+            except ModernFormsConnectionError:
+                if source == SOURCE_ZEROCONF:
+                    return self.async_abort(reason="cannot_connect")
+                return self._show_setup_form({"base": "cannot_connect"})
+            user_input[CONF_MAC] = device.info.mac_address  # type: ignore
+            user_input[CONF_NAME] = device.info.device_name  # type: ignore
+
+        # Check if already configured
+        await self.async_set_unique_id(user_input[CONF_MAC])  # type: ignore
+        self._abort_if_unique_id_configured(updates={CONF_HOST: user_input[CONF_HOST]})  # type: ignore
```

Comment:
I am not happy with these #type: ignore work arounds.   But I am fairly new to python3 typing and dont quite know how to fix.

Im getting the following without them and I dont know a fix for mypy.

`homeassistant/components/modern_forms/config_flow.py:72: error: Unsupported target for indexed assignment ("Optional[Dict[str, Any]]")  [index]`

Reply:
The problem is that `user_input` may be `None`, in which case it's not correct to assign to it.

Reply:
Cool thanks,  cleaned up.